src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala

```diff
@@ -166,49 +166,43 @@ class CassandraSnapshotStore(cfg: Config) extends SnapshotStore with CassandraSt
   }
 
   private def metadata(prepStmt: PreparedStatement, persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Vector[SnapshotMetadata]] = {
-    // TODO the RowIterator is using some blocking, would benefit from a rewrite
-    Future {
-      new RowIterator(prepStmt, persistenceId, criteria.maxSequenceNr).map { row =>
-        SnapshotMetadata(row.getString("persistence_id"), row.getLong("sequence_nr"), row.getLong("timestamp"))
-      }.dropWhile(_.timestamp > criteria.maxTimestamp).toVector
-    }(blockingDispatcher)
-  }
+    def iterate(nr: Long, buf: ListBuffer[SnapshotMetadata]): Future[Vector[SnapshotMetadata]] = {
+      session.execute(prepStmt.bind(persistenceId, nr: JLong)) flatMap { rs =>
+        var rowCount = 0
+        var nextNr = 0l
+
+        def fetchAvailable(rs: ResultSet): Future[Vector[SnapshotMetadata]] = {
+          for (r <- rs.asScala.take(rs.getAvailableWithoutFetching)) {
```

Comment:
Well, why exacly asScala can be scary? rs is an iterator (java iterator), asScala creates just a wrapper for java iterator which looks like scala iterator, which can be infinite.

Consider the following code example:

```
Welcome to Scala version 2.11.7-20150623-000000-1fbce4612c (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_92).
Type in expressions to have them evaluated.
Type :help for more information.

scala>     val javaIter = new java.util.Iterator[String] {
     |       override def next(): String = "hello"
     |       override def hasNext: Boolean = true
     |     }
javaIter: java.util.Iterator[String] = $anon$1@3fb4f649

scala> 

scala> javaIter.next
res0: String = hello

scala> import scala.collection.JavaConverters._
import scala.collection.JavaConverters._

scala> val scalaIter = javaIter.asScala
scalaIter: Iterator[String] = non-empty iterator

scala> scalaIter.next
res1: String = hello

scala> scalaIter.take(5).toList
res2: List[String] = List(hello, hello, hello, hello, hello)
```

It `asScala` loops the whole thing, `javaIter.asScala` invocation should throw OutOfMemory or StackOverflow, but it doesn't. It doesn't loop the whole thing, it just wraps it.